enzyme/Enzyme/MLIR/Implementations/LinalgAutoDiffOpInterfaceImpl.cpp

```diff
@@ -0,0 +1,237 @@
+
+    // get iteration domain
+    AffineMap aMap = newOp.getShapesToLoopsMap();
+    SmallVector<Value> dims;
+    for (OpOperand *input : newOp.getInputOperands()) {
+      auto shape = cast<MemRefType>(input->get().getType()).getShape();
+      for (int i = 0; i < shape.size(); i++) {
+        auto dimI =
+            cacheBuilder.create<arith::ConstantIndexOp>(op->getLoc(), i);
+        // input->get().getParentBlock()->dump();
+        auto dim = cacheBuilder.create<memref::DimOp>(op->getLoc(),
+                                                      input->get(), dimI);
+        dims.push_back(dim);
+      }
+    }
+
+    SmallVector<Value> iterationDomains;
+    std::vector<int64_t> shapes;
+    for (unsigned int i = 0; i < aMap.getNumResults(); i++) {
+      AffineMap subMap = aMap.getSubMap({i});
+      Value domain = cacheBuilder.create<AffineApplyOp>(op->getLoc(), subMap,
+                                                        ValueRange(dims));
+      iterationDomains.push_back(domain);
+      shapes.push_back(-1);
+    }
+
+    SmallVector<Value> inputs, outputs;
+    SmallVector<AffineMap> indexingMaps;
+    SmallVector<StringRef> iteratorTypes{linalgOp.getNumLoops(),
+                                         getParallelIteratorTypeName()};
+
+    for (OpOperand *output : linalgOp.getOutputOperands()) {
+      if (gutils->hasInvertPointer(output->get())) {
+        indexingMaps.push_back(linalgOp.getMatchingIndexingMap(output));
+        Value out = gutils->invertPointerM(output->get(), builder);
+        Value view = invertMemref(out, builder, op->getLoc());
+        outputs.push_back(view);
+      }
+    }
+
+    for (OpOperand *input : linalgOp.getInputOperands()) {
+      if (gutils->hasInvertPointer(input->get())) {
+        indexingMaps.push_back(linalgOp.getMatchingIndexingMap(input));
+        Value inp = gutils->invertPointerM(input->get(), builder);
+        Value view = invertMemref(inp, builder, op->getLoc());
+        inputs.push_back(view);
+      }
+    }
+
+    linalg::GenericOp adjoint = builder.create<linalg::GenericOp>(
+        op->getLoc(), outputs, inputs, indexingMaps, iteratorTypes);
+
+    int numInputs = inputs.size();
+    std::function<buildReturnFunction> buildFuncReturnOp =
+        [numInputs](OpBuilder &builder, Location loc,
+                    SmallVector<Value> retargs) {
+          builder.create<linalg::YieldOp>(
+              loc, ValueRange{retargs}.take_front(numInputs));
+          return;
+        };
+
+    Region *newOpRegion = newOp.getBlock()->getParent();
+    int numInputsNewOp = cast<linalg::GenericOp>(newOp).getInputs().size();
+    Region *adjointRegion = &adjoint.getBodyRegion();
+    int numInputsAdjoint = adjoint.getInputs().size();
+    Location loc = op->getLoc();
+    int numCaches = 0;
+    SmallVector<Value> pushCaches;
+
+    // std::function<std::pair<Value, Value>(Type)> hook = nullptr;
+    std::function<std::pair<Value, Value>(Type)> hook =
```

Comment:
```suggestion
    auto hook =
```
I would rather use `auto` here so that the compiler could avoid creating a `std::function` object.